docs/src/design/stack/precompiles.md

```diff
@@ -0,0 +1,79 @@
+# Precompiles
+
+Precompiles let Miden programs defer expensive computations to the host while still producing
+auditable evidence inside the STARK. This page describes how the VM, host, prover, and verifier
+coordinate to maintain a sequential commitment to every precompile invocation.
+
+## Core data
+
+| Concept | Description |
+| ------- | ----------- |
+| `PrecompileRequest` | Minimal calldata for a precompile, recorded by the host when the event handler runs. It contains exactly the information needed to deterministically recompute the result and the commitment. Requests are included in the proof artifact. |
+| `PrecompileCommitment` | A word pair `(TAG, COMM)` computed by the MASM wrapper, and deterministically recomputable from the corresponding `PrecompileRequest`. `COMM` typically commits to inputs, and may also include outputs for long results; the three free elements in `TAG` carry metadata and/or simple results. Together `(TAG, COMM)` represent the full request (inputs + outputs). |
+| `PrecompileTranscript` | A sequential commitment to all precompile requests. Implemented with an RPO256 sponge; the VM stores only the capacity (4 elements). The verifier reconstructs the same transcript by re‑evaluating requests and their commitments. Finalizing yields a transcript digest. |
+
+## Lifecycle overview
+
+1. **Wrapper emits event** – The MASM wrapper stages inputs (e.g., on stack/memory) and emits the event for the target precompile.
+2. **Host handler runs** – The host executes the event handler, reads required inputs from the current process state, stores a `PrecompileRequest` (raw calldata) for later verification, and pushes the precompile result to the VM via the advice stack.
+3. **Wrapper constructs commitment** – The wrapper pops result(s) from advice, computes `(TAG, COMM)` per the precompile’s convention, and prepares to log the operation.
+4. **`log_precompile` records the commitment** – The wrapper invokes `log_precompile` with `[COMM, TAG, PAD, ...]`. The instruction:
+   - Reads the previous transcript capacity `CAP_PREV` (non‑deterministically via helper registers).
+   - Applies the RPO permutation to `[CAP_PREV, TAG, COMM]`, producing `[CAP_NEXT, R0, R1]`.
+   - Writes `[R1, R0, CAP_NEXT]` back onto the stack; programs typically drop these words immediately.
+5. **Capacity tracking via vtable** – Capacity is tracked inside the VM via the chiplets’ virtual table; the host never tracks capacity. The table always stores the current capacity (the transcript state). On each `log_precompile`:
+   - The previous capacity is removed from the table.
+   - The permutation links `CAP_PREV --[TAG,COMM]--> CAP_NEXT`.
+   - The next capacity is inserted back into the table.
+   This enforces that updates can only occur by applying the permutation.
```

Comment:
I'm still trying to convince myself that the "sibling vtable" + hasher chiplet properly enforce this relationship as currently implemented - will sleep on it and update this comment tomorrow.